### PR TITLE
Add Kill/Death ratio in statistics window

### LINKED ISSUE #6858: Kills/Deaths ratio
It would be interesting to see the kills/deaths ratio in the endgame statistics. Somehow the efficiency of the player.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/5662473-kills-deaths-ratio?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jabbink**
closes #6858

### CODE REVIEW

File: OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs [hunk 1]
```diff
@@ -272,8 +272,11 @@ ScrollItemWidget BasicStats(Player player)
 
 			var stats = player.PlayerActor.TraitOrDefault<PlayerStatistics>();
 			if (stats == null) return template;
-			template.Get<LabelWidget>("KILLS").GetText = () => (stats.UnitsKilled + stats.BuildingsKilled).ToString();
-			template.Get<LabelWidget>("DEATHS").GetText = () => (stats.UnitsDead + stats.BuildingsDead).ToString();
+			var totalKilled = stats.UnitsKilled + stats.BuildingsKilled;
+			var totalDead = stats.UnitsDead + stats.BuildingsDead;
+			template.Get<LabelWidget>("KILLS").GetText = () => totalKilled.ToString();
+			template.Get<LabelWidget>("DEATHS").GetText = () => totalDead.ToString();
+			template.Get<LabelWidget>("KDR").GetText = () => KillDeathRatio(totalKilled, totalDead);
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: I'd prefer `KD_RATIO` here.

File: OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs [hunk 2]
```diff
@@ -304,6 +307,13 @@ string AverageEarnedPerMinute(double earned)
 			return "$" + (world.WorldTick == 0 ? 0 : earned / (world.WorldTick / 1500.0)).ToString("F2");
 		}
 
+		string KillDeathRatio(int killed, int dead)
+		{
+			var kdr = (float)killed;
+			if (dead > 0) kdr /= (float)dead;
```
Comment (Phrohdoh) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]:
```csharp
if (foo)
    DoThing();

OtherMethod();
```

please.

Comment (chrisforbes) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: I'd have probably used max(dead, 1).

Comment (jabbink) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: @Phrodoh I actually had it different before but then I noticed this style further down in the same file (currently mobile so hard to look up).

@chrisforbes thanks, will probably change it that way then.

**Comment by jabbink** [comment]
Renamed Widget to `KD_RATIO` and changed the calculation, which removed the `if` with the missing new line.

File: OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs [hunk 3]
```diff
@@ -272,8 +272,11 @@ ScrollItemWidget BasicStats(Player player)
 
 			var stats = player.PlayerActor.TraitOrDefault<PlayerStatistics>();
 			if (stats == null) return template;
-			template.Get<LabelWidget>("KILLS").GetText = () => (stats.UnitsKilled + stats.BuildingsKilled).ToString();
-			template.Get<LabelWidget>("DEATHS").GetText = () => (stats.UnitsDead + stats.BuildingsDead).ToString();
+			var totalKilled = stats.UnitsKilled + stats.BuildingsKilled;
```
Comment (pchote) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: Pulling these outside the lambda will prevent them from updating while the menu is open.

Comment (jabbink) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: Whoops, completely missed that :fearful:

**Comment by jabbink** [comment]
Restored correct behaviour for the lambda functions again

File: OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs (hunk 2 above)
Comment (jabbink) [inline OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs]: @Phrohdoh https://github.com/delftswa2014/OpenRA/blob/feature/kdr_stat/OpenRA.Mods.RA/Widgets/Logic/ObserverStatsLogic.cs#L316

**Comment by MicroJOo** [comment]
just a simple calling it "K/D" ... would be nicer imo ... 
kd

**Comment by jabbink** [comment]
Everyone agrees? Personally I think that if this changes to `K/D` (or `KDR`), `Actions/min` should also change to `A/m` (or `APM`).

**Comment by Phrohdoh** [comment]
I actually like seeing raw statistics (where applicable) instead of ratios.

**Comment by chrisforbes** [comment]
The raw kills/deaths numbers are still there.

One might argue that `kills - deaths` is more important in an RTS than KDR,
but... this is a number people understand now.

**Comment by jabbink** [comment]
Rebased.

**Comment by obrakmann** [comment]
Looks ok to me, works in both ra and td. :+1:

**Comment by Mailaender** [comment]
Confirmed. Works as promised.

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/d767cac5ac6194f6b8f88ac25959a56d5c8ef3c7...d5b3b0c493e34af24df2fef05cbd4d39c3404964)

**Pull request merged**